cpp/include/raft/spatial/knn/detail/ivf_flat_build.cuh

```diff
@@ -0,0 +1,208 @@
+ * vectors. Note, the total index length is slightly larger than the dataset length, because
+ * each cluster is padded by `WarpSize` elements
+ *
+ * CUDA launch grid:
+ *   X dimension must cover the dataset (n_rows), YZ are not used;
+ *   there are no dependencies between threads, hence no constraints on the block size.
+ *
+ * @tparam T the element type.
+ * @tparam IdxT type of the indices in the source dataset
+ *
+ * @param[in] labels device pointer to the cluster ids for each row [n_rows]
+ * @param[in] list_offsets device pointer to the cluster offsets in the output (index) [n_lists]
+ * @param[in] dataset device poitner to the input data [n_rows, dim]
+ * @param[out] list_data device pointer to the output [index_size, dim]
+ * @param[out] list_index device pointer to the source ids corr. to the output [index_size]
+ * @param[out] list_sizes_ptr device pointer to the cluster sizes [n_lists];
+ *                          it's used as an atomic counter, and must be initialized with zeros.
+ * @param n_rows source length
+ * @param dim the dimensionality of the data
+ * @param veclen size of vectorized loads/stores; must satisfy `dim % veclen == 0`.
+ *
+ */
+template <typename T, typename IdxT>
+__global__ void build_index_kernel(const uint32_t* labels,
+                                   const IdxT* list_offsets,
+                                   const T* dataset,
+                                   T* list_data,
+                                   IdxT* list_index,
+                                   uint32_t* list_sizes_ptr,
+                                   IdxT n_rows,
+                                   uint32_t dim,
+                                   uint32_t veclen)
+{
+  const IdxT i = IdxT(blockDim.x) * IdxT(blockIdx.x) + threadIdx.x;
+  if (i >= n_rows) { return; }
+
+  auto list_id     = labels[i];
+  auto inlist_id   = atomicAdd(list_sizes_ptr + list_id, 1);
+  auto list_offset = list_offsets[list_id];
+
+  // Record the source vector id in the index
+  list_index[list_offset + inlist_id] = i;
+
+  // The data is written in interleaved groups of `index::kGroupSize` vectors
+  using interleaved_group = Pow2<kIndexGroupSize>;
+  auto group_offset       = interleaved_group::roundDown(inlist_id);
+  auto ingroup_id         = interleaved_group::mod(inlist_id) * veclen;
+
+  // Point to the location of the interleaved group of vectors
+  list_data += (list_offset + group_offset) * dim;
+
+  // Point to the source vector
+  dataset += i * dim;
+
+  // Interleave dimensions of the source vector while recording it.
+  // NB: such `veclen` is selected, that `dim % veclen == 0`
+  for (uint32_t l = 0; l < dim; l += veclen) {
+    for (uint32_t j = 0; j < veclen; j++) {
+      list_data[l * kIndexGroupSize + ingroup_id + j] = dataset[l + j];
+    }
+  }
+}
+
+/** See raft::spatial::knn::ivf_flat::build docs */
+template <typename T, typename IdxT>
+inline auto build(const handle_t& handle,
```

Comment:
As discussed offline, we will need to separate the training of the "quantizer" (kmeans in this case) from the ivf.

Reply:
Done via splitting `build` into `build` and `extend`;  `index_params` has now a parameter to tell if we want to add vectors on `build` (default to true).